CIF-404 - Integrate Magento graphQL implementation in cif-magento repository

cjelger opened the pull request:
## Description

Implements the GraphQL CIF API for Magento.

## Related Issue

CIF-404

## How Has This Been Tested?

Added unit and integration tests.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [x] My code passes the code style as defined by ESLint.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes and the overall coverage did not decrease.
- [x] All unit tests pass on CircleCi.
- [x] I ran all tests locally and they pass.

herzog31 commented on src/graphql/graphql.js:
```diff
@@ -0,0 +1,108 @@
+const magentoMapper = require('./magentoMapper');
+
+function main(args) {
+    return introspectionHandler(args, magentoDataHandler);
```
According to the API, `query` is a required parameter. We should use the `InputValidator` to return a proper error message in case the parameter is missing. I also don't see the check to be done inside the `introspectionHandler`.

herzog31 commented on src/graphql/magentoArgsTransformer.js:
```diff
@@ -0,0 +1,156 @@
+    return null;
+}
+
+let transformerFunctions = {
```
Can you add a quick comment to each of those functions to describe the kind of transformation that is done?

herzog31 commented on src/graphql/magentoMapper.js:
```diff
@@ -0,0 +1,294 @@
+ * 
+ * @param {*} config 
+ */
+const configureMapper = (config) => {
```
Can you maybe refactor this whole thing in a class? And please add documentation.

cjelger replied:
That would indeed be nice but unfortunately the "common" GraphQL library expects a simple object with "dumb" mapper functions,  and doesn't get the object itself so one cannot expect that it would for example call `this.mapper()`. That's why I introduced a closure: we can have state but do not rely on a more traditional "object-oriented construct".

herzog31 commented on src/graphql/magentoMapper.js:
```diff
@@ -0,0 +1,294 @@
+        return result;
+    }
+
+    const mapAttributes = (req, data, configurableOpts) => {
```
Can you split this in multiple private methods?

herzog31 commented on test/resources/graphqlQueries.js:
```diff
@@ -0,0 +1,286 @@
+    }
+}`;
+
+const introspectionQuery = `query IntrospectionQuery {
```
Any reason why this is `const` and all the others are not?

cjelger commented:
@herzog31 Thx for the feedback, I addressed all your points, except for the "class vs. closure" comment, where this is not possible without further changes in the GraphQL common library.
